### PR TITLE
A tool to generate profiles.hpp

### PR DESCRIPTION
**Pull request by pleroy**

### CODE REVIEW

File: tools/generate_profiles.cpp [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+  FileDescriptor const* file_descriptor = method_descriptor->file();
+
+  // Process all the messages in that file.
+  for (int i = 0; i < file_descriptor->message_type_count(); ++i) {
+    Descriptor const* message_descriptor = file_descriptor->message_type(i);
+    switch (message_descriptor->extension_count()) {
+      case 0: {
+        // An auxiliary message that is not an extension.  Nothing to do.
+        break;
+    }
```
Comment (eggrobin) [inline tools/generate_profiles.cpp]: badindent.

File: tools/generate_profiles.cpp [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+  }
+
+  // Now mark the fields that have the same name in In and Out as in-out.
+  if (has_in || has_out) {
+    std::sort(field_descriptors.begin(),
+              field_descriptors.end(),
+              [](FieldDescriptor const* left,
+                 FieldDescriptor const* right) {
+      return left->name() < right->name();
+    });
```
Comment (eggrobin) [inline tools/generate_profiles.cpp]: badindent.

File: tools/generate_profiles.cpp [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+    } else if (nested_name == kReturn) {
+      has_return = true;
+    } else {
+      LOG(FATAL) << "Unexpected nested message "
+                 << nested_descriptor->full_name();
+    }
+  }
+
+  // Now mark the fields that have the same name in In and Out as in-out.
+  if (has_in || has_out) {
```
Comment (eggrobin) [inline tools/generate_profiles.cpp]: &&

File: tools/generate_profiles.cpp [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+      cpp_nested_type_[descriptor] += "    int const " +
+                                      size_field_[field_descriptor] + ";\n";
+    }
+  }
+  cpp_nested_type_[descriptor] += "  };\n";
+}
+
+void Generator::ProcessReturn(Descriptor const* descriptor) {
+  CHECK_EQ(1, descriptor->field_count())
+    << descriptor->full_name() << " must have exactly one field";
```
Comment (eggrobin) [inline tools/generate_profiles.cpp]: bad indent.

File: tools/generate_profiles.cpp [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+      break;
+  }
+  FieldOptions const& options = descriptor->options();
+  if (options.HasExtension(serialization::size)) {
+    size_field_[descriptor] = options.GetExtension(serialization::size);
+  }
+}
+
+void Generator::ProcessInOut(Descriptor const* descriptor,
+                  std::vector<FieldDescriptor const*>* field_descriptors) {
```
Comment (eggrobin) [inline tools/generate_profiles.cpp]: badindent.

File: tools/generate_profiles.cpp (hunk 1 above)
Comment (pleroy) [inline tools/generate_profiles.cpp]: Done.

File: tools/generate_profiles.cpp (hunk 2 above)
Comment (pleroy) [inline tools/generate_profiles.cpp]: Done.

File: tools/generate_profiles.cpp (hunk 3 above)
Comment (pleroy) [inline tools/generate_profiles.cpp]: Done.

File: tools/generate_profiles.cpp (hunk 4 above)
Comment (pleroy) [inline tools/generate_profiles.cpp]: Done.

File: tools/generate_profiles.cpp (hunk 5 above)
Comment (pleroy) [inline tools/generate_profiles.cpp]: Done.

**Pull request merged**